homeassistant/components/recorder/migration.py

```diff
@@ -588,21 +589,30 @@ def _apply_update(instance, session, new_version, old_version):  # noqa: C901
         # Add name column to StatisticsMeta
         _add_columns(session, "statistics_meta", ["name VARCHAR(255)"])
     elif new_version == 24:
-        # Delete duplicated statistics
-        delete_duplicates(instance, session)
         # Recreate statistics indices to block duplicated statistics
         _drop_index(connection, "statistics", "ix_statistics_statistic_id_start")
-        _create_index(connection, "statistics", "ix_statistics_statistic_id_start")
         _drop_index(
             connection,
             "statistics_short_term",
             "ix_statistics_short_term_statistic_id_start",
         )
-        _create_index(
-            connection,
-            "statistics_short_term",
-            "ix_statistics_short_term_statistic_id_start",
-        )
+        try:
+            _create_index(connection, "statistics", "ix_statistics_statistic_id_start")
+            _create_index(
+                connection,
+                "statistics_short_term",
+                "ix_statistics_short_term_statistic_id_start",
+            )
+        except DatabaseError:
+            # There may be duplicated statistics entries, delete duplicated statistics
+            # and try again
+            delete_duplicates(instance, session)
+            _create_index(connection, "statistics", "ix_statistics_statistic_id_start")
+            _create_index(
+                connection,
+                "statistics_short_term",
+                "ix_statistics_short_term_statistic_id_start",
+            )
```

Comment:
```suggestion
      for _ in range(2):
          try:
              _create_index(connection, "statistics", "ix_statistics_statistic_id_start")
              _create_index(
                  connection,
                  "statistics_short_term",
                  "ix_statistics_short_term_statistic_id_start",
              )
          except DatabaseError:
              # There may be duplicated statistics entries, delete duplicated statistics
              # and try again
              delete_duplicates(instance, session)
          else:
              break
```

Reply:
with that loop, he need less doublicate code

Reply:
It's pretty nice.
Maybe we should only call `delete_duplicates` the first iteration though?